### PR TITLE
Add Search Panel for Genes

### PR DESCRIPTION
**Pull request by mikaell**
This PR adds a functionality: Search genes in panels

Reference: https://github.com/Clinical-Genomics/scout/issues/3497

* Search is verbatim
* No regex or case conversion is done
* Searches genes.hgnc_id and genes.symbol in Panel collection and returns all results

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the paxa software installed on `hasta`:
 - Log in into hasta: `ssh <USER.NAME>@hasta.scilifelab.se`
 - Activate the staging environment with the command `us` (Use Stage)
 - Run the command `paxa` and follow the instructions. Make sure to specify scout-stage as the resource you request and the server cg-vm1 as server.
1. Log out from the hasta server.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, log out from `cg-vm1` and log in again in the `hasta` server, repeat the `hasta` and `paxa` procedure, which will release the allocated resource (scout-stage) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data2. 

Goto `http://localhost:5000/panels`.

Search for a gene symbol or hgnc id. If present in panel database it should be returned together with panel version.

image

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by mikaell** [comment]
Leaving it here awaiting feedback.

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Better pagination and number of variants returned by queries in `Search SNVs and INDELs` page
 - Refactored and simplified code used for collecting gene variants for `Search SNVs and INDELs` page
 - Fix sidebar panel icons in Case view
+- Ensure genome builds are input to Scout as strings from config files and CLI input
```
Comment (northwestwitch) [inline CHANGELOG.md:20]: Mmm looks like this line belongs to another PR?

File: scout/build/case.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import logging
 from datetime import datetime
 
-from scout.constants import CUSTOM_CASE_REPORTS
+from scout.constants import CUSTOM_CASE_REPORTS, GENOME_BUILD_37, GENOME_BUILD_38
```
Comment (northwestwitch) [inline scout/build/case.py:4]: This is also part or another branch perhaps?

File: scout/commands/load/case.py [hunk 3]
```diff
@@ -109,11 +110,11 @@ def case(
         )
         raise click.Abort()
 
-    if config_data.get("genome_build") not in [37, 38, "37", "38"]:
-        config_data["genome_build"] = int(
+    if config_data.get("genome_build") not in [GENOME_BUILD_37, GENOME_BUILD_38]:
```
Comment (northwestwitch) [inline scout/commands/load/case.py:113]: Same here

File: scout/server/blueprints/panels/views.py [hunk 4]
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
 @templated("panels/panels.html")
 def panels():
     """Show all panels for a user"""
-    if request.method == "POST":  # Edit/create a new panel and redirect to its page
+    LOG.debug("Show all panels for a user")
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:50]:
```suggestion
```

File: scout/server/blueprints/panels/views.py [hunk 5]
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
...
+    LOG.debug("REQUEST: {} {} {}".format(request.args, request.data, request.form))
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:51]:
```suggestion
```

File: scout/server/blueprints/panels/views.py [hunk 6]
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
...
+    if request.method == "POST" and request.form.get("search_for"):
+        # Query db for panels containing the search string
+        search_string = escape(request.form.get("search_for"))
+        panels_found = store.search_panels(search_string)
+
+    # Add new panel
+    if request.method == "POST" and not request.form.get(
+        "search_for"
+    ):  # Edit/create a new panel and redirect to its page
+        LOG.debug("request.method == 'POST'")
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:65]:
```suggestion
```

File: tests/commands/load/test_load_case_cmd.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import tempfile
 
 from scout.commands import cli
+from scout.constants import GENOME_BUILD_37
```
Comment (northwestwitch) [inline tests/commands/load/test_load_case_cmd.py:6]: Ops, here again!

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 8]
```diff
@@ -27,6 +28,48 @@
...
 
+
+
+ 
+<!-- Create a Bootstrap card containing text bar and submit button for searching -->
+{% macro search_gene_card() %}
+<div class="container-float">
+  <div class="card mt-3">
+    <div class="card-body">
+      <form action="{{ url_for('panels.panels') }}" enctype="multipart/form-data" method="POST">
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:39]: enctype="multipart/form-data"  is generally used if the form contains a file upload. Since this is not the case I think it would be best to omit the enctype so it will be the default (application/x-www-form-urlencoded).

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 9]
```diff
@@ -27,6 +28,48 @@
...
+        <div class="row">
+          <div class="col-5">
+            <input type="text" name="search_for" class="form-control" placeholder="Search Gene in Panels" required
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:42]: Have you checked how the gene search is performed on other pages, for instance on the genes page?

The search there is more user friendly than a verbatim search because when you enter 3 letters or more it suggests you the possible genes, by matching the entered letters in gene names and also aliases:

image

What happens in your code:

image

But you don't know if the gene is not found because you spelled it wrong or because it changed official name and it became an alias or because it doesn't exist in Scout.

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 10]
```diff
@@ -27,6 +28,48 @@
...
+                   pattern="^[a-zA-Z0-9_\-]*$" title="Only alphanumeric characters (A-Z+a-z+0-9) and underscores allowed.">
+          </div>
+          <div class="col">
+            <button type="submit" class="btn btn-secondary">Search!</button>
+          </div>
+          <div class="col">
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:48]: Remove this empty col perhaps?

File: scout/parse/models.py [hunk 11]
```diff
@@ -282,7 +282,7 @@ class ScoutLoadConfig(BaseModel):
     gene_fusion_report: Optional[str] = None
     gene_fusion_report_research: Optional[str] = None
     gene_panels: Optional[List[str]] = []
-    genome_build: int = Field(38, alias="human_genome_build")
+    genome_build: str = Field(38, alias="human_genome_build")
```
Comment (northwestwitch) [inline scout/parse/models.py:285]: Here also, this comes from another branch

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 12]
```diff
@@ -27,6 +28,48 @@
...
+          </div>
+      </form>
+
+      <div id="SearchResults" class="card-body">
+        {% if search_result %}
+        <strong> <em>{{search_string}}</em> found in: </strong> <br>
+        {% for name, version in search_result %}
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:55]: Search results are quite difficult to read, especially if a gene is found on many panels and panel versions:

image

Would be nice to apply some formatting to have these results in a table or something

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 13]
```diff
@@ -66,7 +109,7 @@
         </div>
         <div class="row">
           <div class="col text-center">
-            <button type="submit" class="btn btn-secondary">Upload</button>
+            <button type="submit" class="btn btn-secondary">UploadS</button>
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:112]:
```suggestion
            <button type="submit" class="btn btn-secondary">Upload</button>
```

File: scout/server/blueprints/panels/views.py [hunk 14]
```diff
@@ -73,6 +98,8 @@ def panels():
         panel_names=panel_names,
         panel_versions=panel_versions,
         institutes=institutes,
+        search_string=search_string,
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:101]: This string doesn't show in the search form after performing the search. Perhaps avoid passing it to the template?

File: scout/server/blueprints/panels/views.py [hunk 15]
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
...
+    # Add search box and add results if applicable
+    panels_found = []
+    search_string = ""
+    if request.method == "POST" and request.form.get("search_for"):
+        # Query db for panels containing the search string
+        search_string = escape(request.form.get("search_for"))
+        panels_found = store.search_panels(search_string)
+
+    # Add new panel
+    if request.method == "POST" and not request.form.get(
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:62]:
```suggestion
    elif request.method == "POST":
```

**Review by northwestwitch — CHANGES_REQUESTED** [review]
It appears something went wrong with pulls and pushes and this PR is a mix of two distinct branches (search gene in panels and genome builds branch). 

Nice search panel in the gene panels page! The search is working of one enters the exact symbol or a gene, but I think the feature would benefit a a lot from a more user friendly search (with search suggestions), as I wrote in one of my comment.

Also the presentation of the search result might be improved, given that very often multiple panels and panel version present the same gene and as the code is now it shows a long list of  results that are not super easy to read. For instance consider a table or a <ul> element with distinct colors for each distinct gene panel. Or something! 😄

File: CHANGELOG.md (hunk 1 above)
Comment (mikaell) [reply CHANGELOG.md]: Fixed

File: scout/commands/load/case.py (hunk 3 above)
Comment (mikaell) [reply scout/commands/load/case.py:113]: Hmmm… Must have been some mix-up. Fixing now.

File: scout/parse/models.py (hunk 11 above)
Comment (mikaell) [reply scout/parse/models.py]: Fixed.

File: scout/build/case.py (hunk 2 above)
Comment (mikaell) [reply scout/build/case.py]: Fixed

**Comment by mikaell** [comment]
TODOs
- [ ] Make search more user friendly (with search suggestions)
- [ ] Format search results for readability

File: scout/server/blueprints/panels/views.py (hunk 14 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:99]: Ahh, I was looking into the form field, to be precompiled with the term from the previous search!

File: scout/server/blueprints/panels/templates/panels/panels.html (hunk 12 above)
Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html]: Done!

**Comment by northwestwitch** [comment]
Let me know when you are done with the fixes on this one @mikaell, so I'll give it another round!

**Comment by mikaell** [comment]
> Let me know when you are done with the fixes on this one @mikaell, so I'll give it another round!

I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?

**Comment by northwestwitch** [comment]
> I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?

There's a couple of comments of mine that you haven't seen (easy things)

File: tests/commands/load/test_load_case_cmd.py (hunk 7 above)
Comment (mikaell) [reply tests/commands/load/test_load_case_cmd.py]: Fixed

Comment (mikaell) [reply tests/commands/load/test_load_case_cmd.py]: `% grep -rn GENOME_BUILD_37 *` returns nothing now.

File: scout/server/blueprints/panels/templates/panels/panels.html (hunk 9 above)
Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html]: Typeahead is copied from genes-page.

File: scout/server/blueprints/panels/templates/panels/panels.html (hunk 10 above)
Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html:48]: Done!

File: scout/server/blueprints/panels/templates/panels/panels.html (hunk 8 above)
Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html]: Removed!

**Comment by mikaell** [comment]
> > I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?
> 
> There's a couple of comments of mine that you haven't seen (easy things)

Now I think I've fixed all the comments.

**Comment by northwestwitch** [comment]
> Now I think I've fixed all the comments.

Thanks, I'll look at it!

File: scout/adapter/mongo/panel.py [hunk 16]
```diff
@@ -549,23 +549,36 @@ def latest_panels(self, institute_id, include_hidden=False):
...
+        """Return all panels and versions that contain given gene, list is sorted"""
+
+        # Try to cast search string to integer for searching hgnc_id:s.
+        try:
+            search_int = int(search_string)
+        except ValueError as _err:
+            search_int = search_string
+        query = {"$or": [{"genes.hgnc_id": search_int}, {"genes.symbol": search_string}]}
+        result = self.panel_collection.find(query)
+        result_list = [[element["panel_name"], element["version"]] for element in result]
```
Comment (northwestwitch) [inline scout/adapter/mongo/panel.py:583]: What you'll get after this query is this: ['name', 1.0, 'name', 2.0]

But what you need to display results is this: [['name', [1.0, 2.0]]]

And to reformat results you have created 2 new functions (_compile_panel_versions and _compile_panel_versions) below. 

You don't need all that code if you change your query to return just what you need instead. Try using an aggregation:

```
[{
    $match: {
        'genes.hgnc_id': <HGNC-iD>
    }
}, {
    $group: {
        _id: {
            display_name: '$display_name'
        },
        versions: {
            $addToSet: '$version'
        }
    }
}, {
    $sort: {
        display_name: 1
    }
}]
```

That will return results like this:

image

File: scout/server/blueprints/cases/views.py [hunk 17]
```diff
@@ -439,7 +387,14 @@ def phenotypes_actions(institute_id, case_name):
...
     if action == "ADDGENE":
-        hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        try:
+            hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
+                "warning",
+            )
+        # TODO: dont crash
```
Comment (northwestwitch) [inline scout/server/blueprints/cases/views.py:397]:
```suggestion
```

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 18]
```diff
@@ -27,6 +28,68 @@
   </div>
 {% endblock %} <!-- end of content_main -->
 
+
+
+ 
+<!-- Create a Bootstrap card containing text bar and submit button for searching -->
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:34]: This looks really nice now!

File: scout/server/blueprints/cases/views.py [hunk 19]
```diff
@@ -439,7 +387,14 @@ def phenotypes_actions(institute_id, case_name):
             store.remove_phenotype(institute_obj, case_obj, user_obj, case_url, hpo_id)
 
     if action == "ADDGENE":
-        hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        try:
+            hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
```
Comment (northwestwitch) [inline scout/server/blueprints/cases/views.py:394]: In which case does this crash? If I don't type something that it likes I just get this validation error:

image

File: scout/server/blueprints/panels/templates/panels/panels.html [hunk 20]
```diff
@@ -219,5 +287,26 @@ <h5 class="modal-title">Remove gene panel "{{panel.display_name}}"?</h5>
...
+        });
+      }
+
+      $(".typeahead_gene").typeahead({
+        name: 'dynamic_gene_add',
+        source: getNameTerms,
+        minLength: 3,
+      });
+
+      $('select[multiple]').selectpicker({
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/templates/panels/panels.html:305]: I don't think you need this one: you don't have any multiple selects in the form of this page

File: scout/server/blueprints/panels/views.py [hunk 21]
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
...
+        search_string = escape(request.form.get("search_for"))
+        try:
+            hgnc_symbols = parse_raw_gene_ids([search_string])
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
+                "warning",
+            )
+        # TODO: dont crash after error
+        LOG.debug("SEARCH: {}".format(search_string))
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:68]:
```suggestion
```

File: scout/server/blueprints/panels/views.py [hunk 22]
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
...
+        LOG.debug("HGNC: {}".format(hgnc_symbols))
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:69]:
```suggestion
```

File: scout/server/blueprints/panels/views.py [hunk 23]
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
...
+
+        for hgnc_id in hgnc_symbols:
+            panels_found = store.search_panels(hgnc_id)
+            panels_found_compiled.extend(_compile_panel_versions(panels_found))
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:73]: You don't need this if you optimize the query!

File: scout/server/blueprints/panels/views.py [hunk 24]
```diff
@@ -311,3 +351,42 @@ def gene_edit(panel_id, hgnc_id):
                 if panel_value is not None:
                     form_field.process_data(panel_value)
     return dict(panel=panel_obj, form=form, gene=hgnc_gene, panel_gene=panel_gene)
+
+
+def _compile_panel_versions(panels_found):
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:356]: You don't need these 2 functions. Just optimize the query instead

File: scout/server/blueprints/panels/views.py [hunk 25]
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
...
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
+                "warning",
+            )
+        # TODO: dont crash after error
+        LOG.debug("SEARCH: {}".format(search_string))
+        LOG.debug("HGNC: {}".format(hgnc_symbols))
+
+        for hgnc_id in hgnc_symbols:
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:71]: This is not actual any more since you are using the autocompletion no?

If you try searching by more that one symbol the form doesn't submit:

image

image

File: scout/server/blueprints/panels/views.py [hunk 26]
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
...
+            panels_found = store.search_panels(hgnc_id)
+            panels_found_compiled.extend(_compile_panel_versions(panels_found))
+            LOG.debug("FOUND: {}".format(panels_found_compiled))
+    # Add new panel
+    elif request.method == "POST":
+        "search_for"
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:77]: What's this, a comment?

**Review by northwestwitch — CHANGES_REQUESTED** [review]
The looks of the results are much better now and the autocompletion works, but I think it still needs some polishing.

I have a few comments, the most important thing being optimizing the query to avoid a lot of processing of the query results downstream.

File: scout/adapter/mongo/panel.py [hunk 27]
```diff
@@ -549,23 +549,36 @@ def latest_panels(self, institute_id, include_hidden=False):
...
                 {"$match": {"_id": {"$in": panel_ids}}},
                 {"$unwind": "$genes"},
                 {"$group": {"_id": "$genes.hgnc_id"}},
             ]
         )
-        return set(item["_id"] for item in query)
+        return set(item["_id"] for item in query_result)
+
+    def search_panels(self, search_string):
+        """Return all panels and versions that contain given gene, list is sorted"""
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/adapter/mongo/panel.py:574]: Done!

File: scout/server/blueprints/panels/views.py (hunk 25 above)
Comment (mikaell) [reply scout/server/blueprints/panels/views.py:70]: Right now search is only supported for one gene at a time. Though the input check is done in javascript, the backend is not prepared to handle multiple search words separated by commas. 

-Do we/should we support this? 


image

File: scout/server/blueprints/panels/views.py (hunk 26 above)
Comment (mikaell) [reply scout/server/blueprints/panels/views.py:76]: A random string 😃

File: scout/server/blueprints/panels/views.py (hunk 23 above)
Comment (mikaell) [reply scout/server/blueprints/panels/views.py:72]: I don't understand which line you are referring to?

Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:72]: I mean that is you optimize the query you don't need _compile_panel_versions and _compile_panel_versions_aux, the adapter's search_panels function can return the list of panel in the right format

File: scout/server/blueprints/cases/views.py (hunk 17 above)
Comment (mikaell) [reply scout/server/blueprints/cases/views.py:397]: That's always a good reminder, but I'll remove it.

File: scout/server/blueprints/panels/views.py (hunk 23 above)
Comment (mikaell) [reply scout/server/blueprints/panels/views.py:72]: Ok.

But I assume that the a for-look is still necessary as `parse_raw_gene_symbols(raw_symbols_list)` may return a set with several entries?


```
def parse_raw_gene_symbols(raw_symbols_list):
    """Parse list of concatenated gene symbol list for hgnc_symbols from Phenomizer.

    Arguments:
        raw_symbols(list(string)) - e.g. ("POT1 | MUTYH", "POT1 | ATXN1 | ATXN7")

    Returns:
        hgnc_symbols(set(string)) - set of (unique) gene symbols without intervening chars
    """

```

File: scout/server/blueprints/panels/views.py (hunk 24 above)
Comment (mikaell) [reply scout/server/blueprints/panels/views.py:355]: Done!

File: scout/server/blueprints/panels/views.py (hunk 23 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:72]: Yes, I guess so!

File: scout/server/blueprints/panels/templates/panels/panels.html (hunk 20 above)
Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html:305]: I'm not an expert in javascript. But I interpret `$('select[multiple]').selectpicker'` as being a dynamic index, not allowing for multiple selects in a dropdown. Changing the code to `$('select[i]').selectpicker` has no apparent effect on the page's behaviour.

Comment (mikaell) [reply scout/server/blueprints/panels/templates/panels/panels.html:305]: Well, removing it completely changed nothing either. 

Done.

File: scout/server/blueprints/cases/views.py (hunk 19 above)
Comment (mikaell) [reply scout/server/blueprints/cases/views.py:394]: It's possible, I've managed. But don't remember how. 

Anyways, the function `parse_raw_gene_ids()` was extracted from cases/views. Here it had the side-effect of flashing a warning. When moving the functionality to utils/genes.py the side-effect of flashing messages had to be extracted. Therefor to keep the same behaviour the try/except now is outside and kept the same. cases.views.py and panels.views.py will be similar.

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Much improved! I still have a few minor comments.

Take also a look at what is flagged by SonarCloud, perhaps some of these things are easily fixable!

File: scout/adapter/mongo/panel.py [hunk 28]
```diff
@@ -549,23 +549,59 @@ def latest_panels(self, institute_id, include_hidden=False):
...
             [
                 {"$match": {"_id": {"$in": panel_ids}}},
                 {"$unwind": "$genes"},
                 {"$group": {"_id": "$genes.hgnc_id"}},
             ]
         )
-        return set(item["_id"] for item in query)
+        return set(item["_id"] for item in query_result)
+
+    def search_panels(self, search_string):
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/adapter/mongo/panel.py:573]: It's a pretty good API function.

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/adapter/mongo/panel.py:573]: But it's not used, and returns a result very similar to the other, but formatted worse

[the remark this answers is not in the document]
Comment (mikaell) [reply scout/adapter/mongo/panel.py]: OK

File: scout/adapter/mongo/panel.py [hunk 29]
```diff
@@ -549,23 +549,59 @@ def latest_panels(self, institute_id, include_hidden=False):
...
+            self: PanelHandler()
+            search_string:  hgnsc_id (int) to search
+
+        Returns:
+             list: [dict(), dict()]
+        """
+        query = [
+            {"$match": {"genes.hgnc_id": hgnc_id}},
+            {"$group": {"_id": "$display_name", "versions": {"$addToSet": "$version"}}},
+            {"$sort": {"display_name": 1}},
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/adapter/mongo/panel.py]: Fixed.

File: scout/server/blueprints/cases/views.py [hunk 30]
```diff
@@ -439,7 +387,13 @@ def phenotypes_actions(institute_id, case_name):
...
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/server/blueprints/cases/views.py]: I don't trust users 🙂

I've made this crash, but can't reproduce it a work. Maybe I could do it on my home machine because it is so slow and unresponsive on some webpages.

File: scout/adapter/mongo/panel.py [hunk 31]
```diff
@@ -549,23 +549,49 @@ def latest_panels(self, institute_id, include_hidden=False):
...
+        query = [
+            {"$match": {"genes.hgnc_id": hgnc_id}},
+            {
+                "$group": {
+                    "_id": "$display_name",
+                    "versions": {"$addToSet": "$version"},
+                }
+            },
+            {"$unwind": "$versions"},
+            {"$sort": {"_id": 1, "versions": 1}},
```
Comment (northwestwitch) [inline scout/adapter/mongo/panel.py:591]: 🥇

**Review by northwestwitch — APPROVED** [review]
Looks good now! 💯 

image

Feel free to merge. It's going to be a very useful feature!
